File: stackunderflow/etl/backfill.py
"""Backfill orchestrator — convert ``messages`` → ``usage_events`` → marts.

Walks every existing ``messages`` row through the registered normalizers
to materialize ``usage_events`` rows, then refreshes every registered
mart from those events.

Wave 4B fills in the body that Wave 1 skeleton-locked:

* Streams the ``messages`` table (joined to ``sessions`` + ``projects``
  so the normalizer has provider + project_id + session_id without an
  extra round-trip) in chunks of :data:`_CHUNK_SIZE` rows. The user's
  store has ~228K messages — ``fetchall()`` would balloon RAM and
  block the connection for the duration of the run.
* One transaction per chunk so a partial failure leaves a recoverable
  state. Re-runs are idempotent via ``uniq_events_msg`` UNIQUE on
  ``source_message_fk`` (``INSERT OR IGNORE``).
* ``force=True`` first DELETEs from ``usage_events`` + every mart via
  ``rebuild_from_scratch``, **then** runs the normalize pass fresh.
* After events are written, calls
  :func:`stackunderflow.etl.watermark.refresh_all_marts` so every mart
  picks up the new events. (Each mart is watermarked + idempotent on
  its own, so this is the canonical refresh entry-point.)

Logging is intentionally chatty for backfill — emits a progress line
every :data:`_PROGRESS_EVERY_EVENTS` events so the operator knows the
run is alive on a 228K-message store.

See ``docs/specs/etl-architecture.md``.
"""

from __future__ import annotations

import logging
import sqlite3
import time
from collections.abc import Iterator
from dataclasses import dataclass, field

from .marts import all as _all_marts
from .normalize import all as _all_normalizers
from .watermark import refresh_all_marts

_log = logging.getLogger(__name__)

# Chunk size for the streaming SELECT — keeps RAM bounded while still
# amortising the per-chunk transaction overhead. 5,000 rows is roughly
# a few MB of decoded sqlite3.Row objects.
_CHUNK_SIZE = 5_000

# Emit a progress log line every N events inserted (or attempted —
# whichever crosses the boundary first). Keeps the log readable on
# small stores without going silent on large ones.
_PROGRESS_EVERY_EVENTS = 10_000


@dataclass
class BackfillReport:
    """Summary of one ``backfill()`` call.

    Returned to the caller (CLI, API, watcher) so they can render
    progress, log timing, or drive further work. ``marts_refreshed``
    is a copy of the dict returned by
    :func:`stackunderflow.etl.watermark.refresh_all_marts` — empty
    when no mart builders are registered.
    """

    events_inserted: int = 0
    events_skipped_duplicate: int = 0
    marts_refreshed: dict[str, int] = field(default_factory=dict)
    duration_seconds: float = 0.0


def _drop_events_and_marts(conn: sqlite3.Connection) -> None:
    """``force=True`` path: rebuild every mart from scratch + zero out events.

    Schema stays intact (``DELETE``, not ``DROP``). Each mart's
    :meth:`rebuild_from_scratch` deletes its own table; we explicitly
    wipe ``usage_events`` + ``mart_watermark`` here so the next pass
    starts from event id ``1`` against fresh marts.

    Order matters: empty events first (so a mart's
    ``rebuild_from_scratch`` reading partial events doesn't repopulate
    against soon-to-be-deleted rows), then watermarks, then walk every
    registered mart. The mart's own ``DELETE FROM <name>_mart`` call
    inside ``rebuild_from_scratch`` is the canonical clear — looping
    via the registry keeps the orchestrator agnostic of which marts
    exist.
    """
    conn.execute("DELETE FROM usage_events")
    conn.execute("DELETE FROM mart_watermark")
    for mart_cls in _all_marts().values():
        mart_cls().rebuild_from_scratch(conn)


def backfill(
    conn: sqlite3.Connection,
    *,
    force: bool = False,
    progress_callback=None,
) -> BackfillReport:
    """One-shot: convert all existing ``messages`` into ``usage_events``,
    then refresh every mart from the new watermark.

    Default is incremental — already-converted messages are skipped via
    the ``UNIQUE(source_message_fk)`` index (``INSERT OR IGNORE``).

    ``force=True`` first wipes ``usage_events`` + ``mart_watermark`` and
    rebuilds every mart from scratch (via each mart's
    ``rebuild_from_scratch``) before running the normalize pass fresh.

    *progress_callback* is an optional ``Callable[[int, int], None]``
    invoked as ``cb(events_so_far, messages_seen)`` once per chunk.
    Used by the CLI to drive a tqdm bar; library callers leave it
    ``None``.
    """
    start = time.perf_counter()
    report = BackfillReport()

    if force:
        _drop_events_and_marts(conn)

    normalizers = _all_normalizers()
    if not normalizers:
        # No providers registered — nothing to convert. Still call
        # refresh_all_marts so empty marts can finalize their watermarks.
        report.marts_refreshed = refresh_all_marts(conn)
        report.duration_seconds = time.perf_counter() - start
        return report

    inserted, skipped, messages_seen = _run_normalizers(
        conn, normalizers, progress_callback=progress_callback,
    )
    report.events_inserted = inserted
    report.events_skipped_duplicate = skipped

    _log.info(
        "backfill: normalize pass complete — events_inserted=%d "
        "events_skipped_duplicate=%d messages_seen=%d",
        inserted, skipped, messages_seen,
    )

    report.marts_refreshed = refresh_all_marts(conn)

    _log.info(
        "backfill: refresh_all_marts complete — %s",
        " ".join(f"{n}={c}" for n, c in sorted(report.marts_refreshed.items()))
            or "(no marts registered)",
    )

    report.duration_seconds = time.perf_counter() - start
    return report


def _run_normalizers(
    conn: sqlite3.Connection,
    normalizers: dict,
    *,
    progress_callback=None,
) -> tuple[int, int, int]:
    """Stream the messages table and dispatch each row to its normalizer.

    Joins ``messages → sessions → projects`` so the normalizer receives
    a row dict with ``provider`` / ``project_id`` / ``session_id`` for
    free (no per-row lookup). Filters to providers that actually have a
    registered normalizer — there's no point streaming rows we'd skip.

    Each chunk is its own transaction:
      * BEGIN
      * SELECT next chunk
      * Per-row: dispatch to normalizer, ``INSERT OR IGNORE`` each yielded
        event, count inserted vs. skipped via ``cur.rowcount``.
      * COMMIT (or ROLLBACK on error — the caller re-raises).

    A partial failure mid-stream leaves the already-committed chunks in
    place. The next ``backfill()`` resumes via the UNIQUE source_message_fk
    index — already-inserted rows turn into counted skips.

    Returns ``(events_inserted, events_skipped_duplicate, messages_seen)``.
    """
    providers = tuple(sorted(normalizers))
    if not providers:
        return 0, 0, 0

    placeholders = ",".join("?" * len(providers))
    select_sql = f"""
        SELECT m.id            AS id,
               m.session_fk    AS session_fk,
               m.seq           AS seq,
               m.timestamp     AS timestamp,
               m.role          AS role,
               m.model         AS model,
               m.input_tokens  AS input_tokens,
               m.output_tokens AS output_tokens,
               m.cache_read_tokens AS cache_read_tokens,
               m.cache_create_tokens AS cache_create_tokens,
               m.content_text  AS content_text,
               m.tools_json    AS tools_json,
               m.raw_json      AS raw_json,
               m.is_sidechain  AS is_sidechain,
               m.uuid          AS uuid,
               m.parent_uuid   AS parent_uuid,
               m.speed         AS speed,
               s.session_id    AS session_id,
               s.project_id    AS project_id,
               p.provider      AS provider
          FROM messages m
          JOIN sessions s ON s.id = m.session_fk
          JOIN projects p ON p.id = s.project_id
         WHERE p.provider IN ({placeholders})
           AND m.id > ?
         ORDER BY m.id
         LIMIT ?
    """  # noqa: S608 — placeholders are a fixed number of '?', not user input

    inserted = 0
    skipped = 0
    messages_seen = 0
    last_id = 0
    last_progress_log = 0

    while True:
        chunk = _fetch_chunk(conn, select_sql, providers, last_id)
        if not chunk:
            break

        # One transaction per chunk — partial failure leaves prior
        # chunks committed and the next pass resumes via the UNIQUE
        # source_message_fk index.
        conn.execute("BEGIN")
        try:
            for row in chunk:
                msg_row = dict(row)
                last_id = int(msg_row["id"])
                messages_seen += 1

                provider = str(msg_row.get("provider") or "")
                normalizer_cls = normalizers.get(provider)
                if normalizer_cls is None:
                    # Filtered out at the SQL level above, but defensive.
                    continue
                normalizer = normalizer_cls()

                ins, skp = _normalize_and_insert(conn, normalizer, msg_row)
                inserted += ins
                skipped += skp

            conn.execute("COMMIT")
        except Exception:
            conn.execute("ROLLBACK")
            raise

        if progress_callback is not None:
            try:
                progress_callback(inserted, messages_seen)
            except Exception as exc:  # noqa: BLE001 — progress UI must never break ingest
                _log.debug("backfill: progress_callback raised: %s", exc)

        if inserted - last_progress_log >= _PROGRESS_EVERY_EVENTS:
            _log.info(
                "backfill: %d events inserted (%d messages seen)",
                inserted, messages_seen,
            )
            last_progress_log = inserted

        if len(chunk) < _CHUNK_SIZE:
            break

    return inserted, skipped, messages_seen


def _fetch_chunk(
    conn: sqlite3.Connection,
    select_sql: str,
    providers: tuple[str, ...],
    last_id: int,
) -> list:
    """Fetch one chunk of joined messages rows past *last_id*.

    Kept as its own helper so the streaming loop in :func:`_run_normalizers`
    is short and readable, and so tests can patch it to inject a
    failing chunk if needed.
    """
    cur = conn.execute(select_sql, (*providers, last_id, _CHUNK_SIZE))
    return cur.fetchall()


def _normalize_and_insert(
    conn: sqlite3.Connection,
    normalizer,
    msg_row: dict,
) -> tuple[int, int]:
    """Dispatch one ``messages`` row through *normalizer* and persist events.

    Returns ``(inserted, skipped_duplicate)``. Yielded events are
    inserted with ``INSERT OR IGNORE`` against ``uniq_events_msg``;
    when the message has already been converted, ``rowcount`` comes
    back ``0`` and the event counts as a skip.

    Re-uses :func:`stackunderflow.ingest.writer.normalize_and_insert_event`
    so the backfill path and the watcher / ingest writer path share the
    same insert SQL — a single source of truth for the events row shape.
    """
    # Local import — keeps ``ingest`` out of the import path of any
    # caller that just imports the normalize / mart registries.
    from stackunderflow.ingest.writer import normalize_and_insert_event

    inserted = 0
    skipped = 0
    try:
        events = list(normalizer.normalize(msg_row))
    except Exception as exc:  # noqa: BLE001 — never let a poison row stop the run
        _log.debug(
            "backfill: normalizer raised for msg %s: %s",
            msg_row.get("id"), exc,
        )
        return 0, 0

    for ev in events:
        ins, skp = normalize_and_insert_event(conn, msg_row, ev)
        inserted += ins
        skipped += skp
    return inserted, skipped


def _stream_messages_for(
    conn: sqlite3.Connection, providers: tuple[str, ...],
) -> Iterator[dict]:
    """Public-ish helper for tests: stream every joined messages row.

    Returns dict-shaped rows in id order. Mostly used by Wave 4B's tests
    to verify the streaming path doesn't fetchall().
    """
    placeholders = ",".join("?" * len(providers))
    sql = f"""
        SELECT m.id, p.provider
          FROM messages m
          JOIN sessions s ON s.id = m.session_fk
          JOIN projects p ON p.id = s.project_id
         WHERE p.provider IN ({placeholders})
         ORDER BY m.id
    """  # noqa: S608 — fixed-size placeholder list
    for row in conn.execute(sql, providers):
        yield dict(row)

File: rust/crates/stax-etl/src/backfill.rs
//! `etl/backfill.py` (365 ln) — the backfill orchestrator.
//!
//! | Item | Python | Consumed by |
//! |---|---|---|
//! | [`backfill`] | `backfill.backfill` | `routes/etl.rs` (the background task), `stax-cli`'s `etl backfill` |
//! | [`drop_events_and_marts`] | `backfill._drop_events_and_marts` | [`backfill`] |
//!
//! # Why this file is here and not in `stax-server`
//!
//! Batch E ported `etl/backfill.py` and `etl/backfill_jobs.py` into ONE module
//! inside `stax-server`, which was right while the only caller was
//! `POST /api/etl/backfill`. It is not right now: `cli.py`'s `etl backfill`
//! verb calls `stackunderflow.etl.backfill` directly and never touches the job
//! slot, and `stax-cli` may not link `stax-server` (DIV-279). So the two Python
//! files are two Rust modules again — the orchestrator here, in the crate whose
//! charter is `etl/`, and the process-local slot (`backfill_jobs.py`, which
//! genuinely is server state) still in `stax_server::services::etl_backfill`,
//! which re-exports these three names so no route path changed.
//!
//! # The events half is already ported; this is the orchestrator
//!
//! `backfill._run_normalizers` — the streaming keyset walk, the per-chunk
//! transaction, `INSERT OR IGNORE` against `uniq_events_msg`, the WAL checkpoint
//! and the poison-row swallow — is [`crate::normalize::pass::run`], written for
//! RS-3 and pinned by its own tests. Re-transliterating it here would fork it.
//! What that file explicitly did *not* take, by the design note at its top, is
//! exactly what lives here: the `force` wipe, the empty-registry short circuit,
//! and the `refresh_all_marts` call that follows the pass.
//!
//! # `_drop_events_and_marts` is NOT `watermark::rebuild_all_marts`
//!
//! They look interchangeable and are not. `rebuild_all_marts` **stamps a
//! watermark per mart** (`set_watermark(conn, name, max_event_id, now)`);
//! Python's `_drop_events_and_marts` stamps nothing — it `DELETE`s
//! `mart_watermark` wholesale and then calls each builder's
//! `rebuild_from_scratch`, which is `DELETE FROM <name>_mart` + `refresh(conn,
//! 0)` and never touches the watermark table. Reusing `rebuild_all_marts` would
//! leave eight `mart_watermark` rows stamped at the pre-wipe high-water mark
//! *before* the normalize pass re-created the events, so the `refresh_all_marts`
//! that follows would skip every event it had just written and every mart would
//! come back empty. That is the DIV-148-shaped trap in this file: two
//! nearly-identical helpers, one correct.
//! [`tests::the_force_wipe_leaves_no_watermark_behind`] pins it.
//!
//! # The price book seam runs through the CALLER's `NormalizeContext`
//!
//! `stackunderflow etl backfill` (the CLI) runs unprimed — `use_price_book_store`
//! is only ever called by `server.py`'s lifespan — so the CLI verb builds its
//! context from the manifest and the route builds one from `crate::pricing::
//! engine`. Both are correct and they differ; that is DIV-016 / RS-3-082, and
//! it is the reason the context is a parameter here rather than a constant.

use anyhow::Result;
use rusqlite::Connection;

use crate::marts;
use crate::marts::watermark::refresh_all_marts;
use crate::normalize::NormalizeContext;
use crate::normalize::pass;

/// `BackfillReport` — what one [`backfill`] call did.
///
/// Nothing on the wire reads this: `POST /api/etl/backfill` answers `202` with
/// the job id long before the run finishes, and the background worker discards
/// the report exactly as Python's `_run_backfill_in_background` does. It is
/// returned anyway because the CLI verb (wave 8) is the other caller of this
/// function and does render it.
#[derive(Debug, Clone, Default)]
pub struct BackfillReport {
    /// Rows written to `usage_events`.
    pub events_inserted: u64,
    /// Rows `uniq_events_msg` rejected as already-converted.
    pub events_skipped_duplicate: u64,
    /// `messages` rows streamed.
    pub messages_seen: u64,
    /// `{mart_name: events_processed}` in registry order.
    pub marts_refreshed: Vec<(String, i64)>,
    /// `time.perf_counter()` delta across the whole call.
    pub duration_seconds: f64,
}

/// `_drop_events_and_marts` — the `force=True` wipe.
///
/// Order is load-bearing and documented as such in Python: events first, so a
/// mart's `rebuild_from_scratch` cannot repopulate against rows that are about
/// to be deleted; then `mart_watermark`; then every **registered** mart — all
/// eight, not the five `KNOWN_MART_NAMES` the status surface renders.
///
/// # Errors
/// Any SQLite error.
pub fn drop_events_and_marts(conn: &Connection) -> Result<()> {
    conn.execute("DELETE FROM usage_events", [])?;
    conn.execute("DELETE FROM mart_watermark", [])?;
    for mart in marts::all() {
        mart.rebuild_from_scratch(conn)?;
    }
    Ok(())
}

/// `backfill(conn, force=…)` — convert every `messages` row into
/// `usage_events`, then refresh every mart.
///
/// `now` is the `last_refresh_ts` stamp `refresh_all_marts` writes. Python
/// re-reads the clock once **per mart** (`set_watermark` calls
/// `datetime.now(UTC)` itself) where this passes one value for all eight; see
/// the numbered finding in `parity/DIV-e-etl.md`.
///
/// # Errors
/// Any SQLite error from the wipe, the pass, or a mart refresh. Python lets the
/// same errors propagate out of `backfill` into its caller's `except`.
pub fn backfill(
    conn: &Connection,
    ctx: &NormalizeContext,
    force: bool,
    now: &str,
) -> Result<BackfillReport> {
    let started = std::time::Instant::now();
    let mut report = BackfillReport::default();

    if force {
        drop_events_and_marts(conn)?;
    }

    // `if not normalizers:` — the registry is a twenty-entry `const` array here,
    // so this branch cannot be reached. Kept because reaching it in Python skips
    // the pass and *still* calls `refresh_all_marts` ("so empty marts can
    // finalize their watermarks"), and a port that silently dropped the
    // distinction would be one refactor away from dropping the call too.
    if crate::normalize::all().is_empty() {
        report.marts_refreshed = refresh_all_marts(conn, now)?;
        report.duration_seconds = started.elapsed().as_secs_f64();
        return Ok(report);
    }

    let pass_report = pass::run(conn, ctx)?;
    report.events_inserted = pass_report.events_inserted;
    report.events_skipped_duplicate = pass_report.events_skipped_duplicate;
    report.messages_seen = pass_report.messages_seen;

    report.marts_refreshed = refresh_all_marts(conn, now)?;
    report.duration_seconds = started.elapsed().as_secs_f64();
    Ok(report)
}

#[doc(hidden)]
pub mod testdb {
    //! A v030-shaped store for the ETL modules that read `usage_events`.
    //!
    //! `#[doc(hidden)] pub` rather than `#[cfg(test)] pub(crate)`: the status
    //! assembler moved to `stax-reports` and its tests seed the same store, and
    //! a `#[cfg(test)]` module is invisible across a crate boundary. The
    //! alternative was a second copy of this DDL, which is a forked fixture —
    //! the thing `assemble_worktrees_payload` cost the campaign once already.
    //!
    //! The DDL is `stax_etl::marts::testdb::SCHEMA`, which is itself copied from
    //! `stackunderflow/store/migrations/` (v006, v007, v011, v012, v022, v023,
    //! v025, v030) — but that module is `#[cfg(test)] pub(crate)` to *its* crate
    //! and cannot be reached from here. Four columns the mart tests do not need
    //! are added back because the status assembler reads them:
    //! `usage_events.cost_source` is half of its `GROUP BY`, and `account` /
    //! `reasoning_tokens` / `raw_extras` are bound by `insert_event`.

    use rusqlite::Connection;

    pub const SCHEMA: &str = r"
        CREATE TABLE projects (
            id INTEGER PRIMARY KEY, provider TEXT NOT NULL, slug TEXT NOT NULL,
            display_name TEXT NOT NULL, UNIQUE (provider, slug));
        CREATE TABLE sessions (
            id INTEGER PRIMARY KEY, project_id INTEGER NOT NULL, session_id TEXT NOT NULL);
        CREATE TABLE messages (
            id INTEGER PRIMARY KEY, session_fk INTEGER NOT NULL, seq INTEGER NOT NULL,
            timestamp TEXT, role TEXT, model TEXT,
            input_tokens INTEGER DEFAULT 0, output_tokens INTEGER DEFAULT 0,
            cache_read_tokens INTEGER DEFAULT 0, cache_create_tokens INTEGER DEFAULT 0,
            content_text TEXT, tools_json TEXT, raw_json TEXT,
            is_sidechain INTEGER DEFAULT 0, uuid TEXT, parent_uuid TEXT,
            speed TEXT DEFAULT 'standard');
        CREATE TABLE usage_events (
            id INTEGER PRIMARY KEY, source_message_fk INTEGER, project_id INTEGER NOT NULL,
            session_id TEXT NOT NULL, provider TEXT NOT NULL, model TEXT NOT NULL DEFAULT '',
            speed TEXT NOT NULL DEFAULT 'standard', role TEXT NOT NULL DEFAULT 'assistant',
            ts TEXT NOT NULL, day TEXT NOT NULL,
            account TEXT NOT NULL DEFAULT 'default',
            input_tokens INTEGER NOT NULL DEFAULT 0, output_tokens INTEGER NOT NULL DEFAULT 0,
            cache_read_tokens INTEGER NOT NULL DEFAULT 0,
            cache_create_tokens INTEGER NOT NULL DEFAULT 0,
            reasoning_tokens INTEGER NOT NULL DEFAULT 0,
            cost_usd REAL NOT NULL DEFAULT 0.0,
            cost_source TEXT NOT NULL DEFAULT 'rate_card',
            raw_extras TEXT);
        CREATE UNIQUE INDEX uniq_events_msg ON usage_events(source_message_fk);
        CREATE TABLE daily_mart (
            day TEXT NOT NULL, project_id INTEGER NOT NULL, provider TEXT NOT NULL,
            model TEXT NOT NULL DEFAULT '', speed TEXT NOT NULL DEFAULT 'standard',
            input_tokens INTEGER NOT NULL DEFAULT 0, output_tokens INTEGER NOT NULL DEFAULT 0,
            cache_read INTEGER NOT NULL DEFAULT 0, cache_create INTEGER NOT NULL DEFAULT 0,
            message_count INTEGER NOT NULL DEFAULT 0, session_count INTEGER NOT NULL DEFAULT 0,
            cost_usd REAL NOT NULL DEFAULT 0.0,
            PRIMARY KEY (day, project_id, provider, model, speed));
        CREATE TABLE session_mart (
            session_id TEXT PRIMARY KEY, project_id INTEGER NOT NULL, provider TEXT NOT NULL,
            primary_model TEXT, first_ts TEXT NOT NULL, last_ts TEXT NOT NULL,
            message_count INTEGER NOT NULL DEFAULT 0,
            user_message_count INTEGER NOT NULL DEFAULT 0,
            assistant_message_count INTEGER NOT NULL DEFAULT 0,
            input_tokens INTEGER NOT NULL DEFAULT 0, output_tokens INTEGER NOT NULL DEFAULT 0,
            cache_read INTEGER NOT NULL DEFAULT 0, cache_create INTEGER NOT NULL DEFAULT 0,
            cost_usd REAL NOT NULL DEFAULT 0.0, is_one_shot INTEGER NOT NULL DEFAULT 0,
            cwd TEXT);
        CREATE TABLE project_mart (
            project_id INTEGER PRIMARY KEY, provider TEXT NOT NULL, slug TEXT NOT NULL,
            display_name TEXT NOT NULL, first_ts TEXT, last_ts TEXT,
            total_messages INTEGER NOT NULL DEFAULT 0, total_sessions INTEGER NOT NULL DEFAULT 0,
            total_input_tokens INTEGER NOT NULL DEFAULT 0,
            total_output_tokens INTEGER NOT NULL DEFAULT 0,
            total_cache_read INTEGER NOT NULL DEFAULT 0,
            total_cache_create INTEGER NOT NULL DEFAULT 0,
            total_cost_usd REAL NOT NULL DEFAULT 0.0,
            total_user_messages INTEGER NOT NULL DEFAULT 0,
            total_assistant_messages INTEGER NOT NULL DEFAULT 0,
            total_tool_use_messages INTEGER NOT NULL DEFAULT 0,
            total_tool_result_messages INTEGER NOT NULL DEFAULT 0,
            total_commands INTEGER NOT NULL DEFAULT 0,
            total_records INTEGER NOT NULL DEFAULT 0,
            total_errors INTEGER NOT NULL DEFAULT 0,
            errors_by_category TEXT NOT NULL DEFAULT '{}',
            total_cache_read_messages INTEGER NOT NULL DEFAULT 0,
            total_commands_followed_by_interruption INTEGER NOT NULL DEFAULT 0,
            total_command_tools INTEGER NOT NULL DEFAULT 0,
            total_command_steps INTEGER NOT NULL DEFAULT 0);
        CREATE TABLE provider_day_mart (
            day TEXT NOT NULL, provider TEXT NOT NULL, cost_usd REAL NOT NULL DEFAULT 0.0,
            message_count INTEGER NOT NULL DEFAULT 0, session_count INTEGER NOT NULL DEFAULT 0,
            project_count INTEGER NOT NULL DEFAULT 0, PRIMARY KEY (day, provider));
        CREATE TABLE model_day_mart (
            day TEXT NOT NULL, model TEXT NOT NULL, speed TEXT NOT NULL DEFAULT 'standard',
            cost_usd REAL NOT NULL DEFAULT 0.0, input_tokens INTEGER NOT NULL DEFAULT 0,
            output_tokens INTEGER NOT NULL DEFAULT 0, cache_read INTEGER NOT NULL DEFAULT 0,
            cache_create INTEGER NOT NULL DEFAULT 0, message_count INTEGER NOT NULL DEFAULT 0,
            session_count INTEGER NOT NULL DEFAULT 0, PRIMARY KEY (day, model, speed));
        CREATE TABLE tool_mart (
            day TEXT NOT NULL, project_id INTEGER NOT NULL, provider TEXT NOT NULL,
            tool_name TEXT NOT NULL, event_count INTEGER NOT NULL DEFAULT 0,
            cost_usd REAL NOT NULL DEFAULT 0.0, tokens_in INTEGER NOT NULL DEFAULT 0,
            tokens_out INTEGER NOT NULL DEFAULT 0, session_count INTEGER NOT NULL DEFAULT 0,
            calls_total INTEGER NOT NULL DEFAULT 0, cache_read INTEGER NOT NULL DEFAULT 0,
            cache_create INTEGER NOT NULL DEFAULT 0,
            PRIMARY KEY (day, project_id, provider, tool_name));
        CREATE TABLE command_mart (
            day TEXT NOT NULL, project_id INTEGER NOT NULL, command_name TEXT NOT NULL,
            event_count INTEGER NOT NULL DEFAULT 0, cost_usd REAL NOT NULL DEFAULT 0.0,
            tokens_in INTEGER NOT NULL DEFAULT 0, tokens_out INTEGER NOT NULL DEFAULT 0,
            session_count INTEGER NOT NULL DEFAULT 0,
            PRIMARY KEY (day, project_id, command_name));
        CREATE TABLE command_day_mart (
            day TEXT NOT NULL, project_id INTEGER NOT NULL,
            command_count INTEGER NOT NULL DEFAULT 0, PRIMARY KEY (day, project_id));
        CREATE TABLE message_tool_mart (
            id INTEGER PRIMARY KEY, message_id INTEGER NOT NULL, project_id INTEGER NOT NULL,
            session_id TEXT NOT NULL, ts TEXT NOT NULL, day TEXT NOT NULL,
            tool_name TEXT NOT NULL, file_path TEXT, byte_count INTEGER,
            call_index INTEGER NOT NULL, UNIQUE (message_id, tool_name, call_index));
        CREATE TABLE mart_watermark (
            mart_name TEXT PRIMARY KEY, last_event_id INTEGER NOT NULL DEFAULT 0,
            last_refresh_ts TEXT NOT NULL);
        CREATE INDEX idx_message_tool_mart_ts ON message_tool_mart(ts, message_id, tool_name);
        CREATE INDEX idx_projects_slug ON projects(slug);
    ";

    /// An empty v030-shaped store.
    pub fn conn() -> Connection {
        let conn = Connection::open_in_memory().expect("in-memory store");
        conn.execute_batch(SCHEMA).expect("schema");
        conn
    }

    /// One `usage_events` row, with the two `GROUP BY` columns explicit.
    pub fn event(conn: &Connection, id: i64, provider: &str, cost_source: &str) {
        conn.execute(
            "INSERT INTO usage_events (id, source_message_fk, project_id, session_id,
                                       provider, model, ts, day, cost_source)
             VALUES (?, ?, 1, 's1', ?, 'm', '2026-01-01T00:00:00+00:00', '2026-01-01', ?)",
            rusqlite::params![id, id, provider, cost_source],
        )
        .expect("insert event");
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// `_drop_events_and_marts` must leave `mart_watermark` EMPTY. Ported as
    /// `watermark::rebuild_all_marts` instead, eight rows survive at the
    /// pre-wipe high-water mark and the `refresh_all_marts` that follows skips
    /// every event the pass just wrote. This is the trap the module docs name.
    #[test]
    fn the_force_wipe_leaves_no_watermark_behind() {
        let conn = testdb::conn();
        testdb::event(&conn, 1, "claude", "rate_card");
        conn.execute(
            "INSERT INTO daily_mart (day, project_id, provider, model, speed, cost_usd)
             VALUES ('2026-01-01', 1, 'claude', 'm', 'standard', 1.0)",
            [],
        )
        .expect("seed a mart row");
        conn.execute(
            "INSERT INTO mart_watermark (mart_name, last_event_id, last_refresh_ts)
             VALUES ('daily', 999, 't')",
            [],
        )
        .expect("seed a watermark");

        drop_events_and_marts(&conn).expect("wipe");

        let count =
            |sql: &str| -> i64 { conn.query_row(sql, [], |row| row.get(0)).expect("count") };
        assert_eq!(count("SELECT COUNT(*) FROM usage_events"), 0);
        assert_eq!(
            count("SELECT COUNT(*) FROM mart_watermark"),
            0,
            "rebuild_all_marts would have left eight rows here"
        );
        assert_eq!(count("SELECT COUNT(*) FROM daily_mart"), 0);
    }

    /// A `force=false` run over an already-converted store inserts nothing,
    /// counts every row as a skip, and still stamps all eight watermarks.
    #[test]
    fn an_incremental_backfill_is_idempotent_and_still_refreshes_every_mart() {
        let conn = testdb::conn();
        conn.execute_batch(
            "INSERT INTO projects (id, provider, slug, display_name)
                 VALUES (1, 'claude', 'p', 'p');
             INSERT INTO sessions (id, project_id, session_id) VALUES (1, 1, 's1');
             INSERT INTO messages (id, session_fk, seq, timestamp, role, model,
                                   input_tokens, output_tokens)
                 VALUES (1, 1, 0, '2026-01-01T00:00:00+00:00', 'assistant',
                         'claude-sonnet-4-5-20250929', 1000, 100);",
        )
        .expect("seed");

        let manifest = package_dir().join("data").join("models.toml");
        let Ok(ctx) = NormalizeContext::unprimed(&manifest) else {
            // The manifest lives in the Python checkout; skip rather than fail
            // when this test runs from a tree without it.
            return;
        };

        let first = backfill(&conn, &ctx, false, "2026-01-01T00:00:00+00:00").expect("first");
        assert_eq!(first.events_inserted, 1);
        assert_eq!(first.marts_refreshed.len(), 8);

        let second = backfill(&conn, &ctx, false, "2026-01-01T00:00:01+00:00").expect("second");
        assert_eq!(second.events_inserted, 0);
        assert_eq!(second.events_skipped_duplicate, 1);
        assert!(
            second.marts_refreshed.iter().all(|(_, n)| *n == 0),
            "a second refresh must not re-add the same window"
        );
        let marks: i64 = conn
            .query_row("SELECT COUNT(*) FROM mart_watermark", [], |row| row.get(0))
            .expect("count");
        assert_eq!(marks, 8);
    }

    /// `stackunderflow/` in the Python checkout this worktree carries.
    fn package_dir() -> std::path::PathBuf {
        std::path::Path::new(env!("CARGO_MANIFEST_DIR"))
            .join("../../../stackunderflow")
            .canonicalize()
            .unwrap_or_else(|_| std::path::PathBuf::from("/nonexistent"))
    }
}
